File: backend/routers/calculators.py
from schemas import BMIResponse, CalorieResponse, BodyFatResponse

def calculate_bmi(height_cm: float, weight_kg: float) -> BMIResponse:
    """Calculate BMI and return category and ideal weight range"""
    height_m = height_cm / 100
    bmi = weight_kg / (height_m ** 2)
    
    # Determine BMI category
    if bmi < 18.5:
        category = "Underweight"
    elif bmi < 25:
        category = "Normal weight"
    elif bmi < 30:
        category = "Overweight"
    else:
        category = "Obese"
    
    # Calculate ideal weight range (BMI 18.5-24.9)
    ideal_min_weight = 18.5 * (height_m ** 2)
    ideal_max_weight = 24.9 * (height_m ** 2)
    
    return BMIResponse(
        bmi=round(bmi, 1),
        category=category,
        ideal_weight_range={
            "min": round(ideal_min_weight, 1),
            "max": round(ideal_max_weight, 1)
        }
    )

def calculate_calories(age: int, gender: str, height_cm: float, weight_kg: float, 
                      activity_level: str, goal: str) -> CalorieResponse:
    """Calculate BMR, TDEE, and calorie goals"""
    
    # Calculate BMR using Mifflin-St Jeor Equation
    if gender.lower() == "male":
        bmr = 10 * weight_kg + 6.25 * height_cm - 5 * age + 5
    else:  # female
        bmr = 10 * weight_kg + 6.25 * height_cm - 5 * age - 161
    
    # Activity multipliers
    activity_multipliers = {
        "sedentary": 1.2,
        "lightly_active": 1.375,
        "moderately_active": 1.55,
        "very_active": 1.725,
        "extremely_active": 1.9
    }
    
    multiplier = activity_multipliers.get(activity_level.lower(), 1.2)
    tdee = bmr * multiplier
    
    # Calculate calorie goal based on objective
    if goal.lower() == "weight_loss":
        calorie_goal = tdee - 500  # 500 calorie deficit for 1 lb/week loss
    elif goal.lower() == "weight_gain":
        calorie_goal = tdee + 500  # 500 calorie surplus for 1 lb/week gain
    else:  # maintenance
        calorie_goal = tdee
    
    # Macronutrient breakdown (40% carbs, 30% protein, 30% fat)
    protein_calories = calorie_goal * 0.30
    carb_calories = calorie_goal * 0.40
    fat_calories = calorie_goal * 0.30
    
    # Convert to grams (protein: 4 cal/g, carbs: 4 cal/g, fat: 9 cal/g)
    protein_grams = protein_calories / 4
    carb_grams = carb_calories / 4
    fat_grams = fat_calories / 9
    
    return CalorieResponse(
        bmr=round(bmr, 0),
        tdee=round(tdee, 0),
        calorie_goal=round(calorie_goal, 0),
        macronutrient_breakdown={
            "protein": {
                "grams": round(protein_grams, 1),
                "calories": round(protein_calories, 0)
            },
            "carbohydrates": {
                "grams": round(carb_grams, 1),
                "calories": round(carb_calories, 0)
            },
            "fat": {
                "grams": round(fat_grams, 1),
                "calories": round(fat_calories, 0)
            }
        }
    )

def calculate_body_fat(age: int, gender: str, height_cm: float, weight_kg: float,
                      waist_cm: float, neck_cm: float, hip_cm: float = None) -> BodyFatResponse:
    """Calculate body fat percentage using Navy method"""
    
    if gender.lower() == "male":
        # Navy method for men
        body_fat = 495 / (1.0324 - 0.19077 * (waist_cm - neck_cm) / height_cm + 0.15456 * (waist_cm - neck_cm) / height_cm) - 450
    else:
        # Navy method for women (requires hip measurement)
        if hip_cm is None:
            raise ValueError("Hip measurement is required for women")
        
        body_fat = 495 / (1.29579 - 0.35004 * (waist_cm + hip_cm - neck_cm) / height_cm + 0.22100 * (waist_cm + hip_cm - neck_cm) / height_cm) - 450
    
    # Determine body fat category
    if gender.lower() == "male":
        if body_fat < 6:
            category = "Essential fat"
        elif body_fat < 14:
            category = "Athletes"
        elif body_fat < 18:
            category = "Fitness"
        elif body_fat < 25:
            category = "Average"
        else:
            category = "Obese"
        
        ideal_range = {"min": 8, "max": 19}
    else:  # female
        if body_fat < 10:
            category = "Essential fat"
        elif body_fat < 16:
            category = "Athletes"
        elif body_fat < 20:
            category = "Fitness"
        elif body_fat < 25:
            category = "Average"
        else:
            category = "Obese"
        
        ideal_range = {"min": 10, "max": 22}
    
    return BodyFatResponse(
        body_fat_percentage=round(body_fat, 1),
        category=category,
        ideal_range=ideal_range
    )

File: frontend/src/pages/Calculators.js
import React, { useState } from 'react';
import {
  Container,
  Grid,
  Card,
  CardContent,
  Typography,
  TextField,
  Button,
  Box,
  Alert,
  Chip,
  Divider,
} from '@mui/material';
import { Calculate, TrendingUp, Restaurant } from '@mui/icons-material';
import { useAuth } from '../contexts/AuthContext';
import axios from 'axios';

const Calculators = () => {
  const { user } = useAuth();
  const [calculations, setCalculations] = useState({});
  const [loading, setLoading] = useState({});

  const calculateBMI = async () => {
    setLoading(prev => ({ ...prev, bmi: true }));
    try {
      const response = await axios.get('/api/users/calculators/bmi');
      setCalculations(prev => ({ ...prev, bmi: response.data }));
    } catch (error) {
      console.error('BMI calculation error:', error);
    } finally {
      setLoading(prev => ({ ...prev, bmi: false }));
    }
  };

  const calculateCalories = async () => {
    setLoading(prev => ({ ...prev, calories: true }));
    try {
      const response = await axios.get('/api/users/calculators/calories');
      setCalculations(prev => ({ ...prev, calories: response.data }));
    } catch (error) {
      console.error('Calorie calculation error:', error);
    } finally {
      setLoading(prev => ({ ...prev, calories: false }));
    }
  };

  const getBMICategory = (bmi) => {
    if (bmi < 18.5) return { label: 'Underweight', color: 'warning' };
    if (bmi < 25) return { label: 'Normal weight', color: 'success' };
    if (bmi < 30) return { label: 'Overweight', color: 'warning' };
    return { label: 'Obese', color: 'error' };
  };

  return (
    <Container maxWidth="lg" sx={{ mt: 4, mb: 4 }}>
      <Typography variant="h4" component="h1" gutterBottom>
        Health Calculators
      </Typography>
      <Typography variant="h6" color="text.secondary" sx={{ mb: 4 }}>
        Calculate your BMI, calorie needs, and body fat percentage
      </Typography>

      <Grid container spacing={3}>
        {/* BMI Calculator */}
        <Grid item xs={12} md={6}>
          <Card>
            <CardContent>
              <Box sx={{ display: 'flex', alignItems: 'center', mb: 2 }}>
                <Calculate color="primary" sx={{ mr: 1 }} />
                <Typography variant="h6">BMI Calculator</Typography>
              </Box>
              
              <Typography variant="body2" color="text.secondary" sx={{ mb: 3 }}>
                Body Mass Index (BMI) is a measure of body fat based on height and weight.
              </Typography>

              <Box sx={{ mb: 3 }}>
                <Typography variant="body2" sx={{ mb: 1 }}>
                  Height: {user?.height ? `${user.height} cm` : 'Not set in profile'}
                </Typography>
                <Typography variant="body2" sx={{ mb: 2 }}>
                  Weight: {user?.weight ? `${user.weight} kg` : 'Not set in profile'}
                </Typography>
              </Box>

              <Button
                variant="contained"
                fullWidth
                onClick={calculateBMI}
                disabled={loading.bmi || !user?.height || !user?.weight}
                sx={{ mb: 2 }}
              >
                {loading.bmi ? 'Calculating...' : 'Calculate BMI'}
              </Button>

              {calculations.bmi && (
                <Box>
                  <Divider sx={{ my: 2 }} />
                  <Box sx={{ textAlign: 'center' }}>
                    <Typography variant="h3" color="primary" gutterBottom>
                      {calculations.bmi.bmi}
                    </Typography>
                    <Chip
                      label={calculations.bmi.category}
                      color={getBMICategory(calculations.bmi.bmi).color}
                      sx={{ mb: 2 }}
                    />
                    <Typography variant="body2" color="text.secondary">
                      Ideal weight range: {calculations.bmi.ideal_weight_range.min} - {calculations.bmi.ideal_weight_range.max} kg
                    </Typography>
                  </Box>
                </Box>
              )}

              {(!user?.height || !user?.weight) && (
                <Alert severity="info" sx={{ mt: 2 }}>
                  Please complete your profile with height and weight to use this calculator.
                </Alert>
              )}
            </CardContent>
          </Card>
        </Grid>

        {/* Calorie Calculator */}
        <Grid item xs={12} md={6}>
          <Card>
            <CardContent>
              <Box sx={{ display: 'flex', alignItems: 'center', mb: 2 }}>
                <Restaurant color="secondary" sx={{ mr: 1 }} />
                <Typography variant="h6">Calorie Calculator</Typography>
              </Box>
              
              <Typography variant="body2" color="text.secondary" sx={{ mb: 3 }}>
                Calculate your daily calorie needs based on your activity level and goals.
              </Typography>

              <Box sx={{ mb: 3 }}>
                <Typography variant="body2" sx={{ mb: 1 }}>
                  Age: {user?.age || 'Not set'} • Gender: {user?.gender || 'Not set'}
                </Typography>
                <Typography variant="body2" sx={{ mb: 1 }}>
                  Height: {user?.height ? `${user.height} cm` : 'Not set'} • Weight: {user?.weight ? `${user.weight} kg` : 'Not set'}
                </Typography>
                <Typography variant="body2" sx={{ mb: 1 }}>
                  Activity: {user?.activity_level?.replace('_', ' ') || 'Not set'}
                </Typography>
                <Typography variant="body2" sx={{ mb: 2 }}>
                  Goal: {user?.goal?.replace('_', ' ') || 'Not set'}
                </Typography>
              </Box>

              <Button
                variant="contained"
                fullWidth
                onClick={calculateCalories}
                disabled={loading.calories || !user?.age || !user?.gender || !user?.height || !user?.weight || !user?.activity_level || !user?.goal}
                sx={{ mb: 2 }}
              >
                {loading.calories ? 'Calculating...' : 'Calculate Calories'}
              </Button>

              {calculations.calories && (
                <Box>
                  <Divider sx={{ my: 2 }} />
                  <Box sx={{ textAlign: 'center', mb: 2 }}>
                    <Typography variant="h4" color="secondary" gutterBottom>
                      {calculations.calories.calorie_goal} cal/day
                    </Typography>
                    <Typography variant="body2" color="text.secondary">
                      Daily calorie goal
                    </Typography>
                  </Box>
                  
                  <Box sx={{ mb: 2 }}>
                    <Typography variant="subtitle2" gutterBottom>Macronutrient Breakdown:</Typography>
                    <Box sx={{ display: 'flex', justifyContent: 'space-between', mb: 1 }}>
                      <Typography variant="body2">Protein:</Typography>
                      <Typography variant="body2">{calculations.calories.macronutrient_breakdown.protein.grams}g ({calculations.calories.macronutrient_breakdown.protein.calories} cal)</Typography>
                    </Box>
                    <Box sx={{ display: 'flex', justifyContent: 'space-between', mb: 1 }}>
                      <Typography variant="body2">Carbs:</Typography>
                      <Typography variant="body2">{calculations.calories.macronutrient_breakdown.carbohydrates.grams}g ({calculations.calories.macronutrient_breakdown.carbohydrates.calories} cal)</Typography>
                    </Box>
                    <Box sx={{ display: 'flex', justifyContent: 'space-between' }}>
                      <Typography variant="body2">Fat:</Typography>
                      <Typography variant="body2">{calculations.calories.macronutrient_breakdown.fat.grams}g ({calculations.calories.macronutrient_breakdown.fat.calories} cal)</Typography>
                    </Box>
                  </Box>
                </Box>
              )}

              {(!user?.age || !user?.gender || !user?.height || !user?.weight || !user?.activity_level || !user?.goal) && (
                <Alert severity="info" sx={{ mt: 2 }}>
                  Please complete your profile to use this calculator.
                </Alert>
              )}
            </CardContent>
          </Card>
        </Grid>

        {/* Body Fat Calculator */}
        <Grid item xs={12} md={6}>
          <Card>
            <CardContent>
              <Box sx={{ display: 'flex', alignItems: 'center', mb: 2 }}>
                <TrendingUp color="success" sx={{ mr: 1 }} />
                <Typography variant="h6">Body Fat Calculator</Typography>
              </Box>
              
              <Typography variant="body2" color="text.secondary" sx={{ mb: 3 }}>
                Calculate your body fat percentage using the Navy method. Requires waist, neck, and hip measurements.
              </Typography>

              <Alert severity="info" sx={{ mt: 2 }}>
                This calculator requires manual measurements. Please visit your profile to add body measurements.
              </Alert>
            </CardContent>
          </Card>
        </Grid>

        {/* Tips Card */}
        <Grid item xs={12} md={6}>
          <Card>
            <CardContent>
              <Typography variant="h6" gutterBottom>
                💡 Health Tips
              </Typography>
              <Box sx={{ display: 'flex', flexDirection: 'column', gap: 1 }}>
                <Typography variant="body2">
                  • BMI is a screening tool, not a diagnostic tool
                </Typography>
                <Typography variant="body2">
                  • Consult healthcare professionals for personalized advice
                </Typography>
                <Typography variant="body2">
                  • Focus on sustainable lifestyle changes
                </Typography>
                <Typography variant="body2">
                  • Track progress consistently for best results
                </Typography>
              </Box>
            </CardContent>
          </Card>
        </Grid>
      </Grid>
    </Container>
  );
};

export default Calculators;
